### PR TITLE
Cross build for sbt 1.0

### PR DESCRIPTION
**Pull request by jonas**
Bumps dependencies and fixes warnings when compiling with sbt 1.0.

References #150

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -88,9 +87,7 @@ val `scalajs-bundler` =
         commitNextVersion,
         pushChanges,
         releaseStepTask(GhPagesKeys.pushSite in manual)
-      ),
-      scriptedLaunchOpts += "-Dplugin.version=" + version.value,
-      scriptedBufferLog := false
```
Comment (jonas) [inline build.sbt]: Not sure why the top-level project contains settings for the scripted plugins, so I've removed them.

**Comment by jonas** [comment]
I'm seeing a lot of library dependencies warnings, should the sbt-scalajs version be updated in the scripted tests
```
info] [warn] There may be incompatibilities among your library dependencies.
[info] [warn] Here are some of the libraries that were evicted:
[info] [warn] 	* org.scala-js:sbt-scalajs:0.6.18 -> 0.6.19
[info] [warn] Run 'evicted' to see detailed eviction warnings
```

**Comment by sjrd** [comment]
See https://contributors.scala-lang.org/t/a-guide-on-binary-compatibility-need-your-input/1089/19?u=sjrd

**Comment by jonas** [comment]
@sjrd Not sure I understand. Are you saying that we should the sbt-scalajs plugin version as is?

**Comment by sjrd** [comment]
Yes, basically. Or at least that the reporting warning is not a concern and should not itself be the cause to upgrade.

**Comment by jonas** [comment]
Currently working on getting the scripted tests to run for sbt 1.0. I'll need some guidance from the maintainers on how to approach this. Some tests only support sbt 0.13 (for example those depending on sbt-play), other tests we want to run for sbt 1.0 needs to be upgraded to use Scala.js 0.6.19. In addition, we need to remove the `project/build.properties` files from scripted tests that should run for sbt 1.0.

File: build.sbt [hunk 2]
```diff
@@ -135,3 +151,17 @@ lazy val noPublishSettings =
...
+def filterScripted(excludeF: PartialFunction[String, Seq[File => PathFinder]]) = Def.taskDyn {
+  val base = sbtTestDirectory.value
+  val allTests = base * AllPassFilter * AllPassFilter filter { _.isDirectory }
+  val exclude = excludeF((sbtBinaryVersion in pluginCrossBuild).value)
+  val tests = exclude.map(_.apply(base)).foldLeft(allTests)(
+    (remaining, excluded) => remaining --- excluded).get
+  def args = tests.map(Path.relativeTo(base)).flatten.mkString(" ")
+
+  if (tests.nonEmpty)
+    Def.task(scripted.toTask(" " + args).value)
```
Comment (jonas) [inline build.sbt]: This is a super generic way to filter scripted tests based on @xuwei-k's amazing work on https://github.com/sbt/sbt-site/pull/107. If necessary we can simplify it to filter only the last directory, e.g. under `sbt/sbt-{,web}-scalajs-bundle` and include some defaults such as include tests matching `*sbt-${sbtVersion}*`. Let me know what you think/

File: sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+object Compat {
+  type Process = sbt.Process
+  val Process = sbt.Process
+  type ProcessLogger = sbt.ProcessLogger
+
+  // Mimics scala.sys.process.ProcessLogger
+  implicit class ProcessLoggerOps(val self: ProcessLogger) extends AnyVal {
+    def out(msg: String) = self.info(msg)
+  }
+}
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala]: Why not use `scala.sys.Process` everywhere instead of introducing this abstraction over `scala.sys.Process` and `sbt.Process`?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties [hunk 4]
```diff
@@ -1 +0,0 @@
-sbt.version=0.13.15
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties]: Why remove that? I think keeping it (and setting it to `1.0.1`) makes it possible to load the project whatever sbt version is installed.

File: sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala (hunk 3 above)
Comment (jonas) [reply sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala]: I will give it a try.

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties (hunk 4 above)
Comment (jonas) [reply sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties]: If we keep it, it will force the test to only work with that specific sbt version. Most tests I assume we want to run for both versions of sbt. Those we want to fix to a specific sbt version will have to be configured so the CI can filter them out. I've already added a mechanism to do this as pointed out elsewhere.

File: build.sbt (hunk 2 above)
Comment (jonas) [reply build.sbt]: Based on @julienrf 's comment we can make this filtering based on the content of `project/build.properties`. If the file is missing, run for all sbt versions else parse the binary version and run it if it matches `sbtBinaryVersion in pluginCrossBuild`.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala [hunk 5]
```diff
@@ -492,10 +492,11 @@ object ScalaJSBundlerPlugin extends AutoPlugin {
       val installDir = target.value / "scalajs-bundler-jsdom"
       val log = streams.value.log
       val jsdomVersion = (version in installJsdom).value
+      val usingYarn = useYarn.value
       if (!installDir.exists()) {
         log.info(s"Installing jsdom in ${installDir.absolutePath}")
         IO.createDirectory(installDir)
-        install(installDir, useYarn.value, log)(s"jsdom@$jsdomVersion")
+        install(installDir, usingYarn, log)(s"jsdom@$jsdomVersion")
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: I wish https://github.com/sbt/sbt/issues/3299 could be fixed soon so that we don’t have to introduce these redundant identifiers (ie `usingYarn` instead of `useYarn.value`)

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties (hunk 4 above)
Comment (julienrf) [reply sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties]: OK I see! I didn’t notice that the tests were executed with both sbt 1.0 and 0.13. This is great actually :) We should not have to introduce a mechanism testing whether a test should be run with sbt 1.0 or 0.13 only.

File: sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala (hunk 3 above)
Comment (jonas) [reply sbt-scalajs-bundler/src/main/scala-sbt-0.13/scalajsbundler/Compat.scala]: Done in 8f2e318

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala (hunk 5 above)
Comment (jonas) [reply sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: 4b2d2f9 updates to newly released sbt 1.0.2 and allows to remove these workarounds.

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties (hunk 4 above)
Comment (jonas) [reply sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/additonalNpmConfig/project/build.properties]: We need that mechanism for at least `sbt-web-scalajs-bundler/play` until the Play sbt plugin has been published for sbt 1.0 (https://github.com/playframework/playframework/issues/7261). However, based on @sjrd's concern, we should probably also add tests that target the minimum version of sbt 0.13 and sbt 1.0 that scalajs-bundler supports to ensure documentation about those requirements stays up to date.

File: build.sbt (hunk 1 above)
Comment (julienrf) [reply build.sbt]: If I remember correctly that was needed to run the `scripted` command from the root project. I noticed that you use the `runScripted` now instead. What’s the difference between them?

File: build.sbt [hunk 6]
```diff
@@ -135,3 +144,30 @@ lazy val noPublishSettings =
     publish := (),
     publishLocal := ()
   )
+
+def runScriptedTask = Def.taskDyn {
+  val sbtBinVersion = (sbtBinaryVersion in pluginCrossBuild).value
+  val base = sbtTestDirectory.value
+  val testDirectoryFinder = base * AllPassFilter * AllPassFilter filter { _.isDirectory }
```
Comment (julienrf) [inline build.sbt]: Why does that find test directories?

Comment (jonas) [reply build.sbt]: It works similar to the shell glob `.../sbt-test/*/*`

**Comment by jonas** [comment]
OK, squashed and added tests for minimum versions of sbt.

File: build.sbt (hunk 1 above)
Comment (jonas) [reply build.sbt]: Completely missed this comment. I will update CONTRIBUTNG.md to mention the new `runScripted` task. A future improvement could be to run the supported scripted tests as part of the normal test target

```
test in Test := Def.sequential(
   test in Test,
   runScriptedTask
).value,
```

File: appveyor.yml [hunk 7]
```diff
@@ -7,8 +7,8 @@ install:
 build_script:
   - sbt clean compile
 test_script:
-  - sbt test scripted
+  - sbt test runScripted
```
Comment (julienrf) [inline appveyor.yml]: Do you know if it’s possible to have the same matrix thing as we have with travis?

File: build.sbt (hunk 1 above)
Comment (jonas) [reply build.sbt]: Updated CONTRIBUTING.md to mention `runScripted`.

File: appveyor.yml (hunk 7 above)
Comment (jonas) [reply appveyor.yml]: Done in the latest version. Take note of the comment about special handling of `^`. It looks related with https://github.com/appveyor/ci/issues/1500.

**Comment by julienrf** [comment]
This is great work, thanks a lot Jonas!
